### PR TITLE
[BugFix] Fail fast when AutoProcessor degrades to bare tokenizer for multimodal model (#39831)

### LINKED ISSUE #39831: [Bug] GLM-5.3-Flash vision silently broken on main: pinned transformers==5.12.1 lacks glm5_next, AutoProcessor degrades to TokenizersBackend
### Checklist

- [x] I searched related issues but found no solution. (Related: #19048, #36507, #37320 — none cover this exact failure.)
- [x] The bug persists in the latest version (`0.0.0.dev1+gc9a8fba99`, nightly image `20260916`).
- [x] Environment info and a minimal reproducible demo are included below.

### Describe the bug

Serving `zai-org/GLM-5.3-Flash` from a recent `main` build starts successfully and `/model_info` reports `"has_image_understanding": true`, but **every image is silently reduced to exactly 3 tokens** (`<|begin_of_image|>` + 1×`<|image|>` + `<|end_of_image|>`), regardless of resolution or content. The `<|image|>` placeholder is never expanded into visual tokens and no `pixel_values` / `image_grid_thw` are produced, so the model answers from hallucination (e.g., describing a "person wearing a bright yellow jacket" when shown a solid red image). Text-only serving is unaffected. Reproduces via both `/v1/chat/completions` and the native `/generate`.

**Root cause (verified in-container):** the deployment pins `transformers==5.12.1` (per `python/pyproject.toml`), and the transformers **v5.12.1 release does not contain the `glm5_next` module** — it currently only exists on transformers `main`. On 5.12.1, `AutoProcessor.from_pretrained("zai-org/GLM-5.3-Flash", trust_remote_code=True)` **silently returns a `TokenizersBackend`** (a bare tokenizer) instead of raising. SGLang then starts with a tokenizer as the multimodal processor: image *loading* still runs (a corrupt base64 payload correctly returns HTTP 400), but nothing expands placeholders or encodes pixels.

This presumably affects everyone deploying GLM-5.3-Flash from main-derived builds right now. The day-0 `lmsysorg/sglang:glm-5.3-flash` image was built from the #36507 branch with a transformers main snapshot, which is why vision works there.

### Reproduction

Launch (H200 TP8/EP8; shown with secrets redacted):

```bash
python3 -m sglang.launch_server \
  --model-path zai-org/GLM-5.3-Flash --trust-remote-code \
  --tp-size 8 --ep-size 8 \
  --reasoning-parser glm45 --tool-call-parser glm47 \
  --dsa-prefill-backend tilelang --dsa-decode-backend tilelang \
  --kv-cache-dtype bfloat16 --moe-runner-backend deep_gemm \
  --speculative-algorithm EAGLE --speculative-num-steps 5 \
  --speculative-eagle-topk 1 --speculative-num-draft-tokens 6
```

1. **Server-level probe** — `POST /v1/chat/completions` with a `data:image/png;base64,...` image and `temperature: 0`:

   | Request | Expected | Actual |
   |---|---|---|
   | text only (`"Hi"`) | — | `prompt_tokens=13` |
   | + one 256×256 image | ~100+ more tokens | `prompt_tokens=16` (**+3 per image**) |
   | + three 256×256 images | — | `prompt_tokens=22` |

   Answers are hallucinated: solid-red → "Yellow", white square on black → "Triangle", and two *different* images (RGB vs RGBA) yield byte-identical responses.

2. **Framework-level root-cause check** (inside the server container):

   ```python
   import transformers  # 5.12.1
   from transformers import AutoProcessor
   from PIL import Image
   p = AutoProcessor.from_pretrained("zai-org/GLM-5.3-Flash", trust_remote_code=True)
   print(type(p).__name__)                    # TokenizersBackend  <- silent degradation
   out = p(images=[Image.new("RGB", (256, 256))], text="hi", return_tensors="pt")
   print(sorted(out.keys()))                  # ['attention_mask', 'input_ids']  (no pixel_values / image_grid_thw)
   ```

3. **Verified workaround**: install a transformers main snapshot (we used `5.18.0.dev0 @ 22e88feb58affcdd70514db734409c0af8cfa3ed`):

   ```dockerfile
   FROM lmsysorg/sglang:nightly-dev-cu13-20260916-c9a8fba9
   RUN /opt/sglang/bin/pip install --no-cache-dir \
       "transformers @ https://github.com/huggingface/transformers/archive/22e88feb58affcdd70514db734409c0af8cfa3ed.tar.gz"
   ```

   With it, `AutoProcessor` resolves to `Glm5NextProcessor`, a 256×256 image expands to **100 tokens** (grid `1×20×20`, with `pixel_values` / `image_grid_thw` / `mm_token_type_ids`), and the same probes answer correctly: solid-red → "Red", white square → "Square", RGBA red → "Salmon" (alpha composited correctly — thanks for #37320).

   Caveat we hit: transformers `main` currently requires `tokenizers>=0.23.1` while SGLang pins `tokenizers==0.22.2`; we overrode the pin locally and re-validated this workload end-to-end (tokenizer round-trip, chat template, glm45/glm47 parsers, EAGLE speculative serving), but mention it in case the pin is load-bearing elsewhere.

### Suggestions

None of these are blockers for us anymore, but they would save the next deployer a confusing afternoon:

1. Bump the `transformers` pin once a release includes `glm5_next`, or document the required snapshot for GLM-5.3-Flash serving.
2. **Fail loudly (or at least warn) at startup when a multimodal architecture resolves its processor to a bare `PreTrainedTokenizerBase`** — silent degradation turns a packaging/dependency gap into wrong answers at runtime.
3. Optionally, a tiny multimodal smoke test for GLM-5.3-Flash that asserts per-image token expansion (`prompt_tokens` grows by the grid size, not by 3).

Happy to validate any fix on our H200 setup, and thank you for the GLM-5.3-Flash support work in #36507 and the follow-ups — everything else in this build (DSA tilelang paths, deep_gemm KPool fusion, MTP, the glm45/glm47 parsers) has been solid.

### Environment

Output of `python3 -m sglang.check_env` run inside the **unmodified** nightly image (the buggy environment; H200, GPU 1 shown):

```
Python: 3.12.3
CUDA available: True
GPU 0: NVIDIA H200 (Compute Capability 9.0)
CUDA: 13.0 / Driver 565.57.01
PyTorch: 2.13.0+cu130
sglang: 0.0.0.dev1+gc9a8fba99
sgl-kernel: 0.4.7
flashinfer_python: 0.6.18
triton: 3.7.1
transformers: 5.12.1          <-- root cause
tokenizers: 0.22.2 (pinned)
Image: lmsysorg/sglang:nightly-dev-cu13-20260916-c9a8fba9
Deployment: docker compose, 8x H200, TP8/EP8, unified serving
```

### PR DESCRIPTION
**Pull request by andli28**
## Motivation

Fixes #39831. 

When serving a multimodal model, when the architecture is not yet implemented in the current version of the transformers package, `AutoProcessor.from_pretrained` silently degrades (at runtime) and we get a bare TokenizersBackend / PreTrainedTokenizerBase without raising.

As a result, we get hallucinated predictions with only placeholder tokens.

## Modifications

This PR adds an explicit guard in get_processor().

- If a model is multimodal (and not one of the intentional tokenizer exceptions already present in get_processor()), if we return a bare PreTrainedTokenizerBase, we raise a RuntimeError telling the user to update transformers.
- Added unit tests in test_hf_transformers.py to cover both the error trigger and intentional tokenizer exemptions.

## Accuracy Tests

N/A - no change to model outputs.

## Speed Tests and Profiling

N/A

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35145870984](https://github.com/sgl-project/sglang/actions/runs/35145870984)
Latest PR Test (Extra): :x: [Run #35145870792](https://github.com/sgl-project/sglang/actions/runs/35145870792)
Latest PR Test (AMD ROCm 10): :x: [Run #35145870922](https://github.com/sgl-project/sglang/actions/runs/35145870922)

### CODE REVIEW

**Comment by QIN2DIM** [comment]
seriously？👀

**Comment by QIN2DIM** [comment]
Thanks for jumping on this so quickly — we're the reporters of #39831 and can confirm the exact failure mode this guard targets, measured on a live deployment.

Environment: 8×H200, image `lmsysorg/sglang:nightly-dev-cu13-20260916-c9a8fba9` (transformers 5.12.1), GLM-5.3-Flash at TP8/EP8, unified serving.

**On the unpatched build:**
- Every image — regardless of resolution — contributed exactly 3 prompt tokens (`<|begin_of_image|>` + 1×`<|image|>` + `<|end_of_image|>`): text-only prompt = 13 tokens, +1 image = 16, +3 images = 22. Reproduced via both `/v1/chat/completions` and the native `/generate`.
- Answers were hallucinated: solid-red image → "Yellow", white square on black → "Triangle", and two *different* images (RGB vs RGBA) produced byte-identical responses. Meanwhile `/model_info` kept reporting `has_image_understanding: true`.
- In-container root cause confirmed: `AutoProcessor.from_pretrained(...)` returns `TokenizersBackend`, and its output keys are limited to `['attention_mask', 'input_ids']` — precisely the silent degradation this PR fails fast on.

After installing a transformers main snapshot (`Glm5NextProcessor` present), the same server expands a 256×256 image to 100 placeholder tokens (grid `1×20×20`, with `pixel_values` / `image_grid_thw` / `mm_token_type_ids`) and answers correctly ("Red" / "Square" / "Salmon" for RGBA red — alpha compositing OK).

A guard like this would have turned our bisection into an immediate, actionable error message, so +1 from us on landing it. Happy to test-drive the branch on the same H200 setup if that helps review.
